lib/std/meta.zig

```diff
@@ -1321,3 +1322,62 @@ test hasUniqueRepresentation {
     try testing.expect(hasUniqueRepresentation(@Vector(std.simd.suggestVectorLength(u8) orelse 1, u8)));
     try testing.expect(@sizeOf(@Vector(3, u8)) == 3 or !hasUniqueRepresentation(@Vector(3, u8)));
 }
+
+/// Returns whether the given type is comptime-only
+/// (e.g. `type`, `comptime_int`), result is comptime-known.
+pub inline fn comptimeOnly(comptime T: type) bool {
+    return comptime struct {
+        fn impl(comptime U: type, comptime structs: []const type, comptime unions: []const type) bool {
+            return switch (@typeInfo(U)) {
+                .type, .comptime_float, .comptime_int, .@"fn", .enum_literal, .undefined, .null => true,
+                .void, .bool, .noreturn, .int, .float, .error_set, .@"enum", .@"opaque" => false,
+                .array => |a| impl(a.child, structs, unions),
+                .optional => |o| impl(o.child, structs, unions),
+                .error_union => |e| impl(e.payload, structs, unions),
+                .vector => |v| impl(v.child, structs, unions),
+                .@"struct" => |s| std.mem.indexOfScalar(type, structs, U) == null and for (s.fields) |field| {
```

Comment:
The check for structs should skip comptime fields, considering:
```zig
const S = struct {
  comptime a: type = i32,
  b: f32 = 3.14,
};
```
Here, S would _not_ be comptime only.

Reply:
good catch! I added a clause to skip comptime fields, and also #22132.
though, I might remove the handling of comptime-only types from this PR altogether: it seems like the calls to `std.meta.comptimeOnly` eat too much of the comptime evaluation quota, and the fix is to just call the function through a `comptime` block.